FIX: field function name letter case preserved in select with group by

chenqi0805 opened the pull request:
*Issue #, if available:*
#373 

*Description of changes:*
main:
(1) SQLToOperatorConverter.java: add a function to extract raw SQLMethodInvokerExpr method names appearing in select fields
(2) SQLAggregationParser.java: pass in raw method names into .parse and make adapted changes such that columnNode name is populated with the raw expression name.
test:
(3) SQLAggregationParserTest: adapted .parse calls and added a unit test to verify raw function name gets populated in column node.
(4) JdbcTestIT: fixed a test case assertion
(5) TypeInformationIT: added a schema verifying integration test

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

penghuo commented on src/main/java/com/amazon/opendistroforelasticsearch/sql/query/planner/converter/SQLToOperatorConverter.java:
```diff
@@ -76,6 +82,19 @@ public boolean visit(MySqlSelectQueryBlock query) {
         return aggregationParser.getColumnNodes();
     }
 
+    public List<String> extractSelectFunctionNames(List<SQLSelectItem> selectItems) {
```
private?

penghuo commented on src/test/java/com/amazon/opendistroforelasticsearch/sql/esintgtest/JdbcTestIT.java:
```diff
@@ -118,15 +118,16 @@ public void stringOperatorNameCaseInsensitiveTest() {
     @Test
     public void dateFunctionNameCaseInsensitiveTest() {
-        assertEquals(
+        assertTrue(
```
why change to assertTrue?

chenqi0805 replied:
Becase response object comparison does not ignore letter case in schema.name. Thus I used assertTrue together with .equalsIgnoreCase.

penghuo commented on src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/planner/converter/SQLAggregationParserTest.java:
```diff
@@ -55,7 +59,9 @@ public void parseAggWithoutExpressionShouldPass() {
-        parser.parse(mYSqlSelectQueryBlock(sql));
+        MySqlSelectQueryBlock query = mYSqlSelectQueryBlock(sql);
+        List<String> selectMethodNames = prepareRawMethodNames(query, true);
+        parser.parse(query, selectMethodNames);
```
Is it possible to make it as the function to avoid duplicate code in all the test cases?

chenqi0805 replied:
Will take another look.

penghuo commented on src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/planner/converter/SQLAggregationParserTest.java:
```diff
@@ -247,7 +306,28 @@ public void aggregationWithNestedShouldThrowException() {
+        parser.parse(query, selectMethodNames);
+    }
+
+    private List<String> prepareRawMethodNames(MySqlSelectQueryBlock query, Boolean fillWithNull) {
```
Is it the same function used in SQLToOperatorConverter?

chenqi0805 replied:
Same logic. Might be duplicate code with the intent to keep parserTest separate from converter.

penghuo commented on src/main/java/com/amazon/opendistroforelasticsearch/sql/query/planner/converter/SQLToOperatorConverter.java:
```diff
@@ -54,11 +57,14 @@ public SQLToOperatorConverter(Client client, ColumnTypeProvider columnTypeProvid
     @Override
     public boolean visit(MySqlSelectQueryBlock query) {
+        //1. extract function names in select
+        List<String> selectMethodNames = extractSelectFunctionNames(query.getSelectList());
```
Instead of just extracting the selectMethodNames, Is it possible to generate the final select name list regardless the field is method or not?

chenqi0805 replied:
Will take another look.

chenqi0805 replied:
After another look, one way to make the code cleaner is to move the name extraction function and lowercase normalization function into the SQLAggregationParser.parse logic.

penghuo commented on src/main/java/com/amazon/opendistroforelasticsearch/sql/query/planner/converter/SQLToOperatorConverter.java:
```diff
@@ -76,6 +82,19 @@ public boolean visit(MySqlSelectQueryBlock query) {
+            if (selectItemExpr instanceof SQLMethodInvokeExpr) {
+                methodNames.add(((SQLMethodInvokeExpr) selectItemExpr).getMethodName());
+            } else {
+                methodNames.add(null);
```
Avoid to use null value, Consider to use Optional if it is necessary.

chenqi0805 replied:
Will take another look.

chenqi0805 commented:
@penghuo Thanks for the initial review. Here is an updated version of change:

*Description of changes:*
main:
(1) SQLToOperatorConverter.java: refactored rewriteFuncNameToLowerCase into aggregation parser
(2) SQLAggregationParser.java: extract raw SelectItem names for ColumnNode and do lowercase normalization after that.
test:
(3) SQLAggregationParserTest: added a unit test to verify raw function name gets populated in column node.
(4) JdbcTestIT: fixed a test case assertion
(5) TypeInformationIT: added a schema verifying integration test

chenqi0805 commented:
@penghuo @dai-chen Thanks for the review. Could you help me merge as I do not yet have write access.

dai-chen commented:
@chenqi0805 Done. Thanks for your contribution!
